### PR TITLE
py/builtinimport: .__file__ support for external module imports

### PR DESCRIPTION
**Pull request by pmp-p**
extracted from https://github.com/micropython/micropython/pull/4897 following @dpgeorge's request

### CODE REVIEW

**Review by pfalcon — CHANGES_REQUESTED** [review]
This is pretty ugly ifdef spaghetti.

Hard to parse:

~~~
#if MICROPY_PY___FILE__
STATIC void do_execute_raw_code(mp_obj_t module_obj, mp_raw_code_t *raw_code, const char* source_name) {
#else
STATIC void do_execute_raw_code(mp_obj_t module_obj, mp_raw_code_t *raw_code) {
#endif
~~~

is actually

~~~
STATIC void do_execute_raw_code(mp_obj_t module_obj, mp_raw_code_t *raw_code
    #ifdef ..
    , const char* source_name
    #endif
) {
~~~

Elsewhere too.

Various MicroPython codestyle violations and inconsistencies. 

Congrats on the first case of trying to follow MicroPython commit message conventions, but still not there, here should be `py/
builtinimport:` (including no space before ":").

(All the guidelines are where they were all the time: https://github.com/micropython/micropython/blob/master/CODECONVENTIONS.md)

**Comment by pmp-p** [comment]
> is actually
```
STATIC void do_execute_raw_code(mp_obj_t module_obj, mp_raw_code_t *raw_code
    #ifdef ..
    , const char* source_name
    #endif
) {
```
I must admit i stopped reading CODECONVENTIONS after PEP8 so I won't discuss C/C++ readability (though i like Panda3D coding style **much more**)

Please consider helping me more, since i'm unable to write C and Documentation (not english native). 

Also I can only test what i can read and i PR what i (should) test. i still have to learn git as i'm actually copy/pasting from lv_micropython and my wasm port which is certainly not the way to handle safe C editing on the core code.

**Comment by pfalcon** [comment]
> so I won't discuss C/C++ readability (though i like Panda3D coding style much more)

The point here is that if you repeat long constructs, it's: a) much harder for reader to say how much they differ (e.g. just one arg added, or maybe something changes with other args); b) much higher chance that someone later changes just one copy, ignoring the other.

File: py/builtinimport.c [hunk 1]
```diff
@@ -145,11 +145,16 @@ STATIC void do_load_from_lexer(mp_obj_t module_obj, mp_lexer_t *lex) {
...
+#if MICROPY_PY___FILE__
+STATIC void do_execute_raw_code(mp_obj_t module_obj, mp_raw_code_t *raw_code, const char* source_name) {
+#else
 STATIC void do_execute_raw_code(mp_obj_t module_obj, mp_raw_code_t *raw_code) {
+#endif
     #if MICROPY_PY___FILE__
-    // TODO
-    //qstr source_name = lex->source_name;
-    //mp_store_attr(module_obj, MP_QSTR___file__, MP_OBJ_NEW_QSTR(source_name));
+    if (source_name!=NULL){
```
Comment (pfalcon) [inline py/builtinimport.c]: Spaces around "!=", space before "{".

File: py/builtinimport.c [hunk 2]
```diff
@@ -145,11 +145,16 @@ STATIC void do_load_from_lexer(mp_obj_t module_obj, mp_lexer_t *lex) {
...
+        mp_store_attr(module_obj, MP_QSTR___file__, MP_OBJ_NEW_QSTR(qstr_from_str(source_name)));
+    } else
+        mp_store_attr(module_obj, MP_QSTR___file__, MP_OBJ_NEW_QSTR(qstr_from_str("<stdin>") ));
```
Comment (pfalcon) [inline py/builtinimport.c]: `"<stdin>") ))` - stray space in there

**Comment by pmp-p** [comment]
thanks for the help, i hope i got them all.

File: py/builtinimport.c [hunk 3]
```diff
@@ -145,11 +145,16 @@ STATIC void do_load_from_lexer(mp_obj_t module_obj, mp_lexer_t *lex) {
...
+STATIC void do_execute_raw_code(mp_obj_t module_obj, mp_raw_code_t *raw_code
     #if MICROPY_PY___FILE__
-    // TODO
-    //qstr source_name = lex->source_name;
-    //mp_store_attr(module_obj, MP_QSTR___file__, MP_OBJ_NEW_QSTR(source_name));
+    , const char* source_name
+    #endif
+) {
+    #if MICROPY_PY___FILE__
+    if (source_name != NULL) {
```
Comment (dpgeorge) [inline py/builtinimport.c]: Can source_name ever be NULL?

File: py/builtinimport.c [hunk 4]
```diff
@@ -145,11 +145,16 @@ STATIC void do_load_from_lexer(mp_obj_t module_obj, mp_lexer_t *lex) {
 #endif
 
 #if MICROPY_PERSISTENT_CODE_LOAD || MICROPY_MODULE_FROZEN_MPY
-STATIC void do_execute_raw_code(mp_obj_t module_obj, mp_raw_code_t *raw_code) {
+STATIC void do_execute_raw_code(mp_obj_t module_obj, mp_raw_code_t *raw_code
     #if MICROPY_PY___FILE__
```
Comment (dpgeorge) [inline py/builtinimport.c]: These conditionals shouldn't be necessary.  This function is static so the compiler is (usually) smart enough to know not to pass this argument if it's not needed.  Indeed compiling this without this guard gives the same code size for all ports.

**Comment by dpgeorge** [comment]
Thanks for splitting this patch out.

But, what is the reason to add it, did you come across the need for it in some scenario?

File: py/builtinimport.c (hunk 3 above)
Comment (pmp-p) [reply py/builtinimport.c]: dunno, maybe if accessed from a C module ? i crash so much my C code that i'd better be safe than sorry. I leave you sole judge :p

**Comment by pmp-p** [comment]
> did you come across the need for it in some scenario ?

yes for reloading module and sending more debug information to a remote IDE, or worse in case of javascript port locating the right file ressource ( could be other host when CORS is allowed and vfs maps a file to an url )

File: py/builtinimport.c (hunk 4 above)
Comment (pmp-p) [reply py/builtinimport.c]: > (usually)
isn't that a bit risky then ?
on new platform compiler are not always smart at the beginning.. 
also did you test all compilers on win32 platform, it seems there's a zoo in there.

Comment (dpgeorge) [reply py/builtinimport.c]: It's not a problem if the compiler cannot optimise it, it'll just be a little bit larger in code size.  This "trick" is already used in the code base, eg `py/objtype.c:check_for_special_accessors`

File: py/builtinimport.c (hunk 3 above)
Comment (dpgeorge) [reply py/builtinimport.c]: It needs to be understood if it can be NULL (or empty) or not.  I guess it must be non-NULL and non-empty.

Comment (pmp-p) [reply py/builtinimport.c]: if .__file__  attr exists but is None then it can't be printed. But that's probably not possible there

File: py/builtinimport.c (hunk 4 above)
Comment (embeddedt) [reply py/builtinimport.c]: @pmp-p One extra parameter is not going to increase code size by any significant amount, especially since it's never used.

**Comment by dpgeorge** [comment]
Ok, I've merged this in a8e3201b376d931a77e66dc80293c570983f9c7b (with fixes to get esp8266, minimal and bare-arm building, and also added some tests).
